File: egottol/engines/eii/encoders.py
"""Encoding manifold Φ: impulses → embedding z."""

from __future__ import annotations

from typing import List

import numpy as np

from egottol.engines.eii.types import EIIConfig, EIIState, ImpulseEvent


class EncodingManifold:
    """Maps impulse events and probe voltages to fixed-dimension embeddings."""

    MODES = ("rate", "latency", "filter", "population", "continuous")

    def __init__(self, config: EIIConfig):
        self.config = config
        self.alpha = 0.9
        self.tau_lat = 5e-3
        self.tau_m = 2e-3

    def encode(
        self,
        state: EIIState,
        events: List[ImpulseEvent],
        *,
        window_start: float,
    ) -> np.ndarray:
        mode = self.config.encoder_mode
        if mode not in self.MODES:
            raise ValueError(f"Unknown encoder mode: {mode}")

        dispatch = {
            "rate": self._rate,
            "latency": self._latency,
            "filter": self._filter,
            "population": self._population,
            "continuous": self._continuous,
        }
        z = dispatch[mode](state, events, window_start=window_start)
        if self.config.read_noise_std > 0:
            z = z + np.random.normal(0, self.config.read_noise_std, size=z.shape)
        return z.astype(float)

    def _rate(
        self,
        state: EIIState,
        events: List[ImpulseEvent],
        *,
        window_start: float,
    ) -> np.ndarray:
        d = self.config.embedding_dim
        counts = np.zeros(d)
        window_events = [e for e in events if e.t >= window_start]
        for e in window_events:
            ch = e.channel % d
            counts[ch] += 1.0
        return counts / max(self.config.window_T, 1e-12)

    def _latency(
        self,
        state: EIIState,
        events: List[ImpulseEvent],
        *,
        window_start: float,
    ) -> np.ndarray:
        d = self.config.embedding_dim
        z = np.zeros(d)
        window_events = [e for e in events if e.t >= window_start]
        for ch in range(d):
            ch_events = [e for e in window_events if e.channel % d == ch]
            if ch_events:
                t_first = min(e.t for e in ch_events)
                z[ch] = np.exp(-(t_first - self.config.stimulus_time) / self.tau_lat)
        return z

    def _filter(
        self,
        state: EIIState,
        events: List[ImpulseEvent],
        *,
        window_start: float,
    ) -> np.ndarray:
        d = self.config.embedding_dim
        z = state.filter_state.copy()
        if z.shape[0] != d:
            z = np.zeros(d)
        dt = self.config.dt
        decay = np.exp(-dt / max(self.tau_m, dt))
        z *= decay
        for e in events:
            if e.t >= window_start:
                ch = e.channel % d
                z[ch] += e.amplitude / max(self.tau_m, dt)
        state.filter_state = z.copy()
        return z

    def _population(
        self,
        state: EIIState,
        events: List[ImpulseEvent],
        *,
        window_start: float,
    ) -> np.ndarray:
        per_channel = self._rate(state, events, window_start=window_start)
        d = self.config.embedding_dim
        if d == 1:
            return per_channel
        pooled = np.zeros(d)
        chunk = max(1, len(per_channel) // d)
        for i in range(d):
            start = i * chunk
            end = start + chunk if i < d - 1 else len(per_channel)
            pooled[i] = np.mean(per_channel[start:end]) if end > start else 0.0
        return pooled[:d]

    def _continuous(
        self,
        state: EIIState,
        events: List[ImpulseEvent],
        *,
        window_start: float,
    ) -> np.ndarray:
        d = self.config.embedding_dim
        probes = state.voltages[:d]
        if len(probes) < d:
            probes = np.pad(probes, (0, d - len(probes)))
        return probes.astype(float)

File: core/eii/encoders.cpp
#include "encoders.h"

#include <algorithm>
#include <cmath>
#include <numeric>

namespace deepiri {

std::vector<double> EncodingManifold::encode(EIIState& state, const std::vector<ImpulseEvent>& events,
                                              double windowStart) {
    std::vector<double> z;
    switch (config_.encoderMode) {
        case EncoderMode::Rate:
            z = rate(state, events, windowStart);
            break;
        case EncoderMode::Latency:
            z = latency(state, events, windowStart);
            break;
        case EncoderMode::Filter:
            z = filter(state, events, windowStart);
            break;
        case EncoderMode::Population:
            z = population(state, events, windowStart);
            break;
        case EncoderMode::Continuous:
            z = continuous(state, events, windowStart);
            break;
    }
    if (config_.readNoiseStd > 0.0) {
        std::normal_distribution<double> dist(0.0, config_.readNoiseStd);
        for (double& v : z) v += dist(rng_);
    }
    return z;
}

std::vector<double> EncodingManifold::rate(EIIState&, const std::vector<ImpulseEvent>& events, double windowStart) {
    int d = config_.embeddingDim;
    std::vector<double> counts(d, 0.0);
    for (const auto& e : events) {
        if (e.t < windowStart) continue;
        int ch = ((e.channel % d) + d) % d;
        counts[ch] += 1.0;
    }
    double windowT = std::max(config_.windowT, 1e-12);
    for (double& c : counts) c /= windowT;
    return counts;
}

std::vector<double> EncodingManifold::latency(EIIState&, const std::vector<ImpulseEvent>& events,
                                               double windowStart) {
    int d = config_.embeddingDim;
    std::vector<double> z(d, 0.0);
    for (int ch = 0; ch < d; ++ch) {
        bool found = false;
        double tFirst = 0.0;
        for (const auto& e : events) {
            if (e.t < windowStart) continue;
            int c = ((e.channel % d) + d) % d;
            if (c != ch) continue;
            if (!found || e.t < tFirst) {
                tFirst = e.t;
                found = true;
            }
        }
        if (found) {
            z[ch] = std::exp(-(tFirst - config_.stimulusTime) / tauLat);
        }
    }
    return z;
}

std::vector<double> EncodingManifold::filter(EIIState& state, const std::vector<ImpulseEvent>& events,
                                              double windowStart) {
    int d = config_.embeddingDim;
    std::vector<double> z = state.filterState;
    if (static_cast<int>(z.size()) != d) {
        z.assign(d, 0.0);
    }
    double dt = config_.dt;
    double decay = std::exp(-dt / std::max(tauM, dt));
    for (double& v : z) v *= decay;
    for (const auto& e : events) {
        if (e.t < windowStart) continue;
        int ch = ((e.channel % d) + d) % d;
        z[ch] += e.amplitude / std::max(tauM, dt);
    }
    state.filterState = z;
    return z;
}

std::vector<double> EncodingManifold::population(EIIState& state, const std::vector<ImpulseEvent>& events,
                                                   double windowStart) {
    std::vector<double> perChannel = rate(state, events, windowStart);
    int d = config_.embeddingDim;
    if (d == 1) return perChannel;
    std::vector<double> pooled(d, 0.0);
    size_t chunk = std::max<size_t>(1, perChannel.size() / static_cast<size_t>(d));
    for (int i = 0; i < d; ++i) {
        size_t start = static_cast<size_t>(i) * chunk;
        size_t end = (i < d - 1) ? start + chunk : perChannel.size();
        if (end > start && start < perChannel.size()) {
            end = std::min(end, perChannel.size());
            double sum = std::accumulate(perChannel.begin() + start, perChannel.begin() + end, 0.0);
            pooled[i] = sum / static_cast<double>(end - start);
        } else {
            pooled[i] = 0.0;
        }
    }
    return pooled;
}

std::vector<double> EncodingManifold::continuous(EIIState& state, const std::vector<ImpulseEvent>&, double) {
    int d = config_.embeddingDim;
    std::vector<double> probes(d, 0.0);
    size_t n = std::min(state.voltages.size(), static_cast<size_t>(d));
    for (size_t i = 0; i < n; ++i) probes[i] = state.voltages[i];
    return probes;
}

}  // namespace deepiri
